widget/data_listener.go

```diff
@@ -0,0 +1,128 @@
+package widget
+
+import (
+	"fmt"
+	"reflect"
+	"strconv"
+
+	"fyne.io/fyne/dataapi"
+)
+
+// DataSetter interface
+type DataSetter interface {
+	SetFromData(dataapi.DataItem)
+}
+
+// DataMapSetter interface
+type DataMapSetter interface {
+	SetFromData(dataapi.DataMap)
+}
+
+// DataSourceSetter interface, for things that can be set from a source
+type DataSourceSetter interface {
+	SetFromSource(source dataapi.DataSource)
+}
+
+// DataListener is a base struct that all DataAPI aware widgets can embed
+type DataListener struct {
+	data       dataapi.DataItem
+	source     dataapi.DataSource
+	onChanged  reflect.Value
+	listenerID int
+}
+
+// StringToInter objects can get an int index from a string (Radio / Select, etc)
+type StringToInter interface {
+	AsInt(string) int
+}
+
+// This gets a bit sticky, because the OnChange handlers on each widget are not methods, but attribs
+// so some reflection magic is needed to see if the object has the appropriate OnChange handler,
+// and set it accordingly
+func getFunction(obj interface{}, name string) (reflect.Value, bool) {
+	f := reflect.ValueOf(obj).Elem().FieldByName(name)
+	return f, f.IsValid() && f.CanSet()
+}
+
+// Bind will Bind this widget to the given DataItem
```

Comment:
I'm not sure about this approach of using reflection to detect, access, and update a function. Can't this be done with interfaces and multiple (type-specific) Bind functions? ie. BindString, BindInt, BindFloat64

Reply:
agree.  Not 100% sure how yet, need to find time for a good session to test the ideas.